File: app/src/main/java/org/gittner/osmbugs/parser/MapdustParser.java
package org.gittner.osmbugs.parser;

import org.gittner.osmbugs.bugs.MapdustBug;
import org.gittner.osmbugs.common.Comment;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class MapdustParser
{
    public static ArrayList<MapdustBug> parse(InputStream stream)
    {
        ArrayList<MapdustBug> bugs = new ArrayList<>();
        BufferedReader reader;
        try
        {
            reader = new BufferedReader(new InputStreamReader(stream, "iso-8859-1"));
            String line = reader.readLine();

            /* Little Tweak since this was sometimes missing in Answer. Only on Android not on Desktop */
            if (!line.endsWith("\"}}"))
            {
                line += "\"}}";
            }
            JSONObject json = new JSONObject(line);
            JSONArray bugArray = json.getJSONArray("features");
            for (int i = 0; i != bugArray.length(); ++i)
            {
                JSONObject bug = bugArray.getJSONObject(i);
                long id = bug.getLong("id");

                JSONObject geometry = bug.getJSONObject("geometry");
                double lon = geometry.getJSONArray("coordinates").getDouble(0);
                double lat = geometry.getJSONArray("coordinates").getDouble(1);

                JSONObject property = bug.getJSONObject("properties");
                MapdustBug.STATE state;
                switch (property.getInt("status"))
                {
                    case 2:
                        state = MapdustBug.STATE.CLOSED;
                        break;
                    case 3:
                        state = MapdustBug.STATE.IGNORED;
                        break;
                    default:
                        state = MapdustBug.STATE.OPEN;
                        break;
                }

                ArrayList<Comment> comments = new ArrayList<>();
                JSONArray commentArray = property.getJSONArray("comments");
                for (int n = 0; n != commentArray.length(); ++n)
                {
                    JSONObject comment = commentArray.getJSONObject(n);
                    comments.add(new Comment(comment.getString("comment")));
                }

                String description = property.getString("description");
                String type_const = property.getString("type");

                int typeInt;
                switch (type_const)
                {
                    case "wrong_turn":
                        typeInt = MapdustBug.WRONG_TURN;
                        break;
                    case "bad_routing":
                        typeInt = MapdustBug.BAD_ROUTING;
                        break;
                    case "oneway_road":
                        typeInt = MapdustBug.ONEWAY_ROAD;
                        break;
                    case "blocked_street":
                        typeInt = MapdustBug.BLOCKED_STREET;
                        break;
                    case "missing_street":
                        typeInt = MapdustBug.MISSING_STREET;
                        break;
                    case "wrong_roundabout":
                        typeInt = MapdustBug.ROUNDABOUT_ISSUE;
                        break;
                    case "missing_speedlimit":
                        typeInt = MapdustBug.MISSING_SPEED_INFO;
                        break;
                    default:
                        typeInt = MapdustBug.OTHER;
                        break;
                }

                bugs.add(new MapdustBug(lat, lon, id, typeInt, description, comments, state));
            }
        }
        catch (IOException e)
        {
            e.printStackTrace();
            return new ArrayList<>();
        }
        catch (JSONException e)
        {
            e.printStackTrace();
            return new ArrayList<>();
        }
        return bugs;
    }


    public static ArrayList<Comment> parseSingleBugForComments(InputStream stream)
    {
        ArrayList<Comment> comments = new ArrayList<>();
        BufferedReader reader;
        try
        {
            reader = new BufferedReader(new InputStreamReader(stream, "iso-8859-1"));
            String line = reader.readLine();

            /*
             * Little Tweak since this was sometimes missing in Answer. Only on Android not on
             * Desktop
             */
            if (!line.endsWith("\"}}"))
            {
                line += "\"}}";
            }

            JSONObject json = new JSONObject(line);

            JSONObject properties = json.getJSONObject("properties");

            JSONArray commentsArray = properties.getJSONArray("comments");
            for (int i = 0; i != commentsArray.length(); ++i)
            {
                comments.add(new Comment(
                        commentsArray.getJSONObject(i).getString("comment"),
                        commentsArray.getJSONObject(i).getString("nickname")));
            }
        }
        catch (IOException e)
        {
            e.printStackTrace();
            return new ArrayList<>();
        }
        catch (JSONException e)
        {
            e.printStackTrace();
            return new ArrayList<>();
        }

        return comments;
    }
}

File: app/src/main/java/org/gittner/osmbugs/mapdust/MapdustParser.kt
package org.gittner.osmbugs.mapdust

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.joda.time.format.ISODateTimeFormat
import org.json.JSONObject
import org.osmdroid.util.GeoPoint


class MapdustParser {
    companion object {
        suspend fun parse(data: String): ArrayList<MapdustError> = withContext(Dispatchers.Default) {
            val errors = ArrayList<MapdustError>()

            val json = JSONObject(data)
            val bugArray = json.getJSONArray("features")
            for (i in 0 until bugArray.length()) {
                val error = bugArray.getJSONObject(i)

                errors.add(parseError(error))
            }

            errors
        }

        suspend fun parseError(data: String): MapdustError = withContext(Dispatchers.Default) {
            parseError(JSONObject(data))
        }

        private fun parseError(error: JSONObject): MapdustError {
            val formatter = ISODateTimeFormat.dateTimeParser()

            val id = error.getLong("id")

            val geometry = error.getJSONObject("geometry")
            val lon = geometry.getJSONArray("coordinates").getDouble(0)
            val lat = geometry.getJSONArray("coordinates").getDouble(1)

            val properties = error.getJSONObject("properties")

            val user = properties.getString("nickname")

            val state = when (properties.getInt("status")) {
                2 -> MapdustError.STATE.CLOSED
                3 -> MapdustError.STATE.IGNORED
                else -> MapdustError.STATE.OPEN
            }

            val comments: ArrayList<MapdustError.MapdustComment> = ArrayList()
            val commentArray = properties.getJSONArray("comments")
            for (n in 0 until commentArray.length()) {
                comments.add(
                    MapdustError.MapdustComment(
                        commentArray.getJSONObject(n).getString("comment"),
                        commentArray.getJSONObject(n).getString("nickname")
                    )
                )
            }

            val description = properties.getString("description")

            val type_const = properties.getString("type")

            val creationDate = formatter.parseDateTime(properties.getString("date_created"))

            val type = when (type_const) {
                "wrong_turn" -> MapdustError.ERROR_TYPE.WRONG_TURN
                "bad_routing" -> MapdustError.ERROR_TYPE.BAD_ROUTING
                "oneway_road" -> MapdustError.ERROR_TYPE.ONEWAY_ROAD
                "blocked_street" -> MapdustError.ERROR_TYPE.BLOCKED_STREET
                "missing_street" -> MapdustError.ERROR_TYPE.MISSING_STREET
                "wrong_roundabout" -> MapdustError.ERROR_TYPE.ROUNDABOUT_ISSUE
                "missing_speedlimit" -> MapdustError.ERROR_TYPE.MISSING_SPEED_INFO
                else -> MapdustError.ERROR_TYPE.OTHER
            }

            val mapdustError = MapdustError(GeoPoint(lat, lon), id, creationDate, user, type, description, state)

            mapdustError.Comments = comments

            return mapdustError
        }

        suspend fun parseAddBug(data: String): Long = withContext(Dispatchers.Default) {
            val json = JSONObject(data)

            json.getLong("id")
        }
    }
}
